### PR TITLE
Causality in Conv2d

### PR DESCRIPTION
**Pull request by fpaissan**

### CODE REVIEW

**Comment by mravanelli** [comment]
@ycemsubakan, are you fine with that? I think it is important to double check to make sure everything is fine.

**Comment by ycemsubakan** [comment]
> @ycemsubakan, are you fine with that? I think it is important to double check to make sure everything is fine.

I ll try to take another look..

**Comment by fpaissan** [comment]
Hey @ycemsubakan @mravanelli, I think I found an issue... [this colab](https://colab.research.google.com/drive/1gKWX_R5Cqgr3J9trzYqAAUDXYpUay3wf?usp=sharing) demos the causality of the conv2d. However, it works on the wrong axis... (channel axis) will solve this right away and push the updated code with the updated colab to demo and review the PR.

**Comment by ycemsubakan** [comment]
Thanks for the PR @fpaissan! 

I have made few commit to make sure that the padding works with both `skip_transpose=False` and `skip_transpose=True`. Basically, the first spatial axis (the first axis on the kernel) is treated as the time axis, and therefore the padding is applied on the first spatial axis in both cases. 

You can test that the method works correctly with this code snippet: 
```
    skip_transpose = True
    conv2d = Conv2d(1, kernel_size=(3, 1), padding='causal', in_channels=1, skip_transpose=skip_transpose, bias=False)
    conv2d.conv.weight = torch.nn.Parameter(torch.ones(conv2d.conv.weight.shape))

    if skip_transpose:
        inp = torch.arange(1, 10).unsqueeze(1).repeat(1, 2).unsqueeze(0).unsqueeze(0).float()
    else:
        inp = torch.arange(1, 10).unsqueeze(1).repeat(1, 2).unsqueeze(0).unsqueeze(-1).float().permute(0, 2, 1, 3)

    result = conv2d(inp)

    print('input is', inp)
    print(inp.shape)
    print('kernel is: ', conv2d.conv.weight)
    print('result is', result)
    print(result.shape)

```
@mravanelli, feel free to test with the code snippet above.

**Review by ycemsubakan — APPROVED** [review]
I think we are ready to merge. But feel free to double check with the code snippet I have put above.

**Comment by mravanelli** [comment]
Thank you @fpaissan and @ycemsubakan!

**Review by mravanelli — APPROVED** [review]
